include/openssl/hpke.h

```diff
@@ -0,0 +1,499 @@
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param ciphertext is a random value of the appropriate length for ciphertext
+ * @param ciphertext_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ *
+ * If suite_in is provided that will be used (if supported). If
+ * suite_in is NULL, a random suite (from those supported) will
+ * be selected. In all cases the output pub and cipher values
+ * will be appropriate random values for the selected suite.
+ */
+int OSSL_HPKE_get_grease_value(OSSL_LIB_CTX *libctx, const char *propq,
+                               OSSL_HPKE_SUITE *suite_in,
+                               OSSL_HPKE_SUITE *suite,
+                               unsigned char *pub,
+                               size_t *pub_len,
+                               unsigned char *ciphertext,
+                               size_t ciphertext_len);
+
+/**
+ * @brief map a string to a HPKE suite
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ *
+ * An example good string is "x25519,hkdf-sha256,aes-128-gcm"
+ * Symbols are #define'd for the relevant labels, e.g.
+ * OSSL_HPKE_KEMSTR_X25519. Numeric (decimal or hex) values with
+ * the relevant IANA codepoint values from RFC9180 may be used,
+ * e.g., "0x20,1,1" represents the same suite as the first
+ * example.
+ */
+int OSSL_HPKE_str2suite(const char *str, OSSL_HPKE_SUITE *suite);
+
+/**
+ * @brief tell the caller how big the cipertext will be
+ * @param suite is the suite to be used
+ * @param clearlen is the length of plaintext
+ * @return the length of the related ciphertext or zero on error
+ *
+ * AEAD algorithms add a tag for data authentication.
+ * Those are almost always, but not always, 16 octets
+ * long, and who know what'll be true in the future.
+ * So this function allows a caller to find out how
+ * much data expansion they'll see with a given
+ * suite.
+ */
+size_t OSSL_HPKE_get_ciphertext_size(OSSL_HPKE_SUITE suite, size_t clearlen);
+
+/**
+ * @brief tell the caller how big the public value ``enc`` will be
+ * @param suite is the suite to be used
+ * @return size of public encap or zero on error
+ *
+ * AEAD algorithms add a tag for data authentication.
+ * Those are almost always, but not always, 16 octets
+ * long, and who know what'll be true in the future.
+ * So this function allows a caller to find out how
+ * much data expansion they'll see with a given
+ * suite.
+ */
+size_t OSSL_HPKE_get_public_encap_size(OSSL_HPKE_SUITE suite);
+
+/**
+ * @brief recommend an IKM size in octets for a given suite
+ * @param suite is the suite to be used
+ * @return the recommended size or zero on error
+ *
+ * Today, this really only uses the KEM to recommend
+ * the number of random octets to use based on the
+ * size of a private value. In future, it could also
+ * factor in e.g. the AEAD.
+ */
+size_t OSSL_HPKE_recommend_ikmelen(OSSL_HPKE_SUITE suite);
```

Comment:
_get_recommended_ikmelen() ?
I would leave this API out unless you have a real need for it..

Reply:
hmm, disagree - if one is using an IKM  the length should ideally be commensurate with the private key size which is known to the implementation and not necessarily the caller, not having this would create an excuse for using ikm="0123456" ;-) I get that providing this might encourage use of an IKM where that's not needed but we can't prevent that

Reply:
Is the IKME only really needed for testing?

Reply:
no, the IKM is not only needed for testing, the deterministic key generation scheme which requires the IKM is part of the RFC so applications or protocols may chose to use that if they wish and hence it ought be supported by the API. I agree that deterministic key gen is a dangerous implement and have a comment on that in the POD file. That could well be improved and comments are welcome. Note that not providing APIs enabling use of IKM wouldn't prevent applications doing exactly that anyway, and those who like to leak keys (e.g. for "entreprise" purposes) are certainly capable of figuring that out.

Reply:
_get_recommended_ikmelen() ?

Reply:
sure, it's a bit of a long name but did that